micro/request.go

```diff
@@ -25,16 +25,41 @@ type (
 	// Request represents service request available in the service handler.
 	// It exposes methods to respond to the request, as well as
 	// getting the request data and headers.
-	Request struct {
-		msg          *nats.Msg
-		respondError error
+	Request interface {
+		// Respond sends the response for the request.
+		// Additional headers can be passed using [WithHeaders] option.
+		Respond([]byte, ...RespondOpt) error
+
+		// RespondJSON marshals the given response value and responds to the request.
+		// Additional headers can be passed using [WithHeaders] option.
+		RespondJSON(interface{}, ...RespondOpt) error
+
+		// Error prepares and publishes error response from a handler.
+		// A response error should be set containing an error code and description.
+		// Optionally, data can be set as response payload.
+		Error(code, description string, data []byte, opts ...RespondOpt) error
+
+		// Data returns request data.
+		Data() []byte
+
+		// Headers returns request headers.
+		Headers() Headers
+
+		// Subject returns underlying NATS message subject.
+		Subject() string
 	}
 
 	// RequestHandler is a function used as a Handler for a service.
-	RequestHandler func(*Request)
+	RequestHandler func(Request)
```

Comment:
Handler should probably be interface based here too a la http.Handler, that would make it convenient to pass around types that can handle requests, think something like this:
```go
// micro.Handler
type Handler func(Request)
func (fn RequestHandler) Handle(req Request) {
  fn(req)
}
```

Reply:
Very good idea, thanks!

Reply:
added